geode-core/src/test/java/org/apache/geode/internal/cache/ClientServerTransactionDUnitTest.java

```diff
@@ -892,6 +892,82 @@ public Object call() throws Exception {
     }
   }
 
+  @Test
+  public void keySetFromClientRegionWillGetKeysFromServerIfTX() throws Exception {
+    Host host = Host.getHost(0);
+    VM server = host.getVM(0);
+    VM client = host.getVM(1);
+    VM client2 = host.getVM(2);
+    int port1 = createRegionsAndStartServer(server, false);
+    createClientRegionAndPopulateData(client, port1, false);
+    createClientRegion(client2, port1, false);
+
+    client2.invoke(new SerializableRunnable("verify client region with tx") {
+      public void run() throws Exception {
+        TXManagerImpl mgr = getGemfireCache().getTxManager();
+        Region<OrderId, Order> orderRegion = getCache().getRegion(ORDER);
+        LocalRegion lr = (LocalRegion) orderRegion;
+        assertEquals(DataPolicy.NORMAL, orderRegion.getAttributes().getDataPolicy());
+        CustId custId = new CustId(1);
+        OrderId orderId = new OrderId(1, custId);
+        Set setWithoutTX = orderRegion.keySet();
+        Iterator iterWithoutTX = setWithoutTX.iterator();
+        if (!iterWithoutTX.hasNext()) {
+          org.apache.geode.test.dunit.LogWriterUtils.getLogWriter().info("No keys in region");
+        } else {
+          org.apache.geode.test.dunit.LogWriterUtils.getLogWriter()
+              .info("Region size:" + orderRegion.size());
+        }
+        // without tx, the local region with NORMAL policy will get nothing from server
```

Comment:
This comment is wrong. By default, region.keySet does not go to server without tx. You should use region.keySetOnServer to get server keys.

The keySet with tx is behaved like keySetOnServer -- as all operations in a client transaction are performed on server. And those operations are applied to client cache (if it not empty) once commit is successful on the server.